### PR TITLE
fix bug in splitSqlScript

### LINKED ISSUE #3316: MySQL Script execution failed if LOOP statement at PROCEDURE used
I'm trying to test PROCEDURE at MySQL testcontainers. This is an example from https://dev.mysql.com/doc/refman/5.7/en/statement-labels.html

Add this to testcontainers-java/modules/mysql/src/test/resources/somepath/init_mysql.sql 
```
CREATE PROCEDURE doiterate(p1 INT)
  BEGIN
    label1: LOOP
      SET p1 = p1 + 1;
      IF p1 < 10 THEN
        ITERATE label1;
      END IF;
      LEAVE label1;
    END LOOP label1;
  END;

CREATE PROCEDURE dowhile()
  BEGIN
    DECLARE v1 INT DEFAULT 5;
    WHILE v1 > 0 DO
      SET v1 = v1 - 1;
    END WHILE;
  END;

CREATE PROCEDURE dorepeat(p1 INT)
  BEGIN
    SET @x = 0;
    REPEAT
      SET @x = @x + 1;
    UNTIL @x > p1 END REPEAT;
  END;
```
And run org.testcontainers.jdbc.mysql.MySQLJDBCDriverTest
It wil be failed
```
Caused by: org.testcontainers.ext.ScriptUtils$ScriptStatementFailedException: Script execution failed (somepath/init_mysql.sql:8): CREATE PROCEDURE doiterate(p1 INT) BEGIN
    label1: LOOP
      SET p1 = p1 + 1;
      IF p1 < 10 THEN
        ITERATE label1;
      END IF
	at org.testcontainers.jdbc.JdbcDatabaseDelegate.execute(JdbcDatabaseDelegate.java:49)
	at org.testcontainers.delegate.AbstractDatabaseDelegate.execute(AbstractDatabaseDelegate.java:34)
	at org.testcontainers.ext.ScriptUtils.executeDatabaseScript(ScriptUtils.java:362)
	... 43 more
Caused by: java.sql.SQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 6
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:764)
	at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:648)
	at org.testcontainers.jdbc.JdbcDatabaseDelegate.execute(JdbcDatabaseDelegate.java:42)
```
Than i run
```
$ docker run --name=mysql1 -d mysql/mysql-server:5.7.22
```
It works fine
```
mysql> DELIMITER $$
mysql> CREATE PROCEDURE doiterate(p1 INT)
    -> BEGIN
    ->     label1: LOOP
    ->         SET p1 = p1 + 1;
    ->         IF p1 < 10 THEN ITERATE label1; END IF;
    ->         LEAVE label1;
    ->     END LOOP label1;
    -> END$$
Query OK, 0 rows affected (0.00 sec)

mysql> CREATE PROCEDURE dowhile()
    -> BEGIN
    ->     DECLARE v1 INT DEFAULT 5;
    ->     WHILE v1 > 0 DO
    ->       SET v1 = v1 - 1;
    ->     END WHILE;
    -> END$$
Query OK, 0 rows affected (0.00 sec)

mysql> DELIMITER ;
mysql> CALL dowhile();
Query OK, 0 rows affected (0.01 sec)
```
Can you explain me what is wrong?
Or why it not supported.

p.s. PROCEDURE whithout LOOP statements works fine.

### LINKED ISSUE #4441: TC_INITSCRIPT can't handle PostgreSQL dollar quoted strings
A simple program like this:

```java
public static void main(String[] args) throws SQLException {
    Properties properties = new Properties();
    properties.setProperty("username", "postgres");
    properties.setProperty("password", "postgres");

    new ContainerDatabaseDriver().connect(
        "jdbc:tc:postgresql:13:///sakila?TC_TMPFS=/testtmpfs:rw&TC_INITSCRIPT=file:c:/temp/dollar-quoted.sql",
        properties
    );
}
```

Where the `dollar-quoted.sql` file contains:

```sql
CREATE FUNCTION f ()
RETURNS INT
AS $$
BEGIN
    RETURN 1;
END;
$$ LANGUAGE plpgsql;
```

Fails with this exception:

```
Exception in thread "main" org.testcontainers.ext.ScriptUtils$UncategorizedScriptException: Failed to execute database script from resource [CREATE FUNCTION f ()
RETURNS INT
AS $$
BEGIN
    RETURN 1;
END;
$$ LANGUAGE plpgsql;
]
	at org.testcontainers.ext.ScriptUtils.executeDatabaseScript(ScriptUtils.java:375)
	at org.testcontainers.ext.ScriptUtils.executeDatabaseScript(ScriptUtils.java:313)
	at org.testcontainers.jdbc.ContainerDatabaseDriver.runInitScriptIfRequired(ContainerDatabaseDriver.java:196)
	at org.testcontainers.jdbc.ContainerDatabaseDriver.connect(ContainerDatabaseDriver.java:132)
	at org.jooq.example.test.containers.TestContainersTest.main(TestContainersTest.java:34)
Caused by: org.testcontainers.ext.ScriptUtils$ScriptStatementFailedException: Script execution failed (file:c:/temp/dollar-quoted.sql:1): CREATE FUNCTION f () RETURNS INT AS $$ BEGIN
    RETURN 1;
END
	at org.testcontainers.jdbc.JdbcDatabaseDelegate.execute(JdbcDatabaseDelegate.java:49)
	at org.testcontainers.delegate.AbstractDatabaseDelegate.execute(AbstractDatabaseDelegate.java:34)
	at org.testcontainers.ext.ScriptUtils.executeDatabaseScript(ScriptUtils.java:362)
	... 4 more
Caused by: org.postgresql.util.PSQLException: Unterminated dollar quote started at position 36 in SQL CREATE FUNCTION f () RETURNS INT AS $$ BEGIN
    RETURN 1;
END. Expected terminating $$
	at org.postgresql.core.Parser.checkParsePosition(Parser.java:1328)
	at org.postgresql.core.Parser.parseSql(Parser.java:1227)
	at org.postgresql.core.Parser.replaceProcessing(Parser.java:1179)
	at org.postgresql.core.CachedQueryCreateAction.create(CachedQueryCreateAction.java:43)
	at org.postgresql.core.QueryExecutorBase.createQueryByKey(QueryExecutorBase.java:337)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:300)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:284)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:279)
	at org.testcontainers.jdbc.JdbcDatabaseDelegate.execute(JdbcDatabaseDelegate.java:42)
	... 6 more
```

It seems that the logic borrowed from Spring's `ScriptUtils` doesn't handle real world SQL scripts very well? I can think of numerous other issues that aren't currently supported, including e.g. MySQL's `DELIMITER` statement, etc.

### PR DESCRIPTION
**Pull request by inponomarev**
Fix  #3316
Fix #4441

# Problem

Current version of `ScriptUtils.splitSqlScript` tries to take into account `BEGIN...END` blocks. However, it treats any `END` as an end of the block, even if it's `END IF` or `END LOOP`, so that the following one-statement script

```sql
BEGIN
     rec_loop: LOOP
         FETCH blah;
         IF something_wrong THEN LEAVE rec_loop; END IF;
         do_something_else;
     END LOOP;
END;
```

Is incorrectly  split into following "statements":

```sql 
BEGIN
    rec_loop: LOOP
        FETCH blah;
        IF something_wrong THEN LEAVE rec_loop; END IF
-----------------------------------------------------------------
       do_something_else
-----------------------------------------------------------------
   END LOOP
-----------------------------------------------------------------
END
```

This is also broken for the case when a special statement separator is used (like `GO` in MSSQL tradition, `/` in Oracle tradition or `@` in DB2).

The following statement is not split over `@` sign as when it comes to the separator, the algorithm counted more `END`s than `BEGIN`s and and since the balance is not zero, it is not splitting the code:

```sql
BEGIN
    rec_loop: LOOP
        FETCH blah;
        IF something_wrong THEN LEAVE rec_loop; END IF;
        do_something_else;
    END LOOP;
END;
@
CALL something();
@
``` 

# Solution

Apparently, the "true end of block" must be `END` immediately followed by a separator (with probably whitespace and/or comments preceding the separator).

The current implementation of `splitSqlScript` is poorly written and it's difficult to extend "syntax rules". So I have completely rewritten it in a style of a standard parser: 

1 "Lexical analysis"  is perfromed in `ScriptScanner` class. This class retrieves next id, next comment, next string literal etc.
2 "Syntactic rules" are implemented in `ScriptSplitter` whis is just a recursive procedural parser.

### CODE REVIEW

**Comment by kiview** [comment]
Hey @inponomarev, thanks for your PR, but unfortunately we will have to close it.

`ScriptUtils` has been mostly taken from the Spring-JDBC project as a convenient feature and we agree that it can't support all use cases and all SQL syntax variants. We also agree with your assessment regarding the difficulty to extend and maintain it. But that's precisely the reason why we don't want to invest further in this feature or refactoring it extensively.

Instead, we recommend to use the mechanisms provided by the vendor specific images, which mostly means using `withCopyFileToContainer()` and a specific path for init scripts. I am pretty sure it should also work for the use case you have in mind. 

In the future, we'd suggest to reach out to us first via Slack or GH discussions, to avoid unnecessary work on features that won't get merged.

**Comment by kiview** [comment]
Re-opened after private discussion in Slack.

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptScanner.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            }
+            end += blockCommentEndDelimiter.length();
+            currentMatch = blockCommentStartDelimiter + script.substring(offset, end);
+            offset = end;
+            return true;
+        }
+        return false;
+    }
+
+    boolean matchesDollarQuotedString() {
```
Comment (kiview) [inline modules/database-commons/src/main/java/org/testcontainers/ext/ScriptScanner.java:102]:
```suggestion
    private boolean matchesDollarQuotedString() {
```

Any reason for package-private?

**Review by kiview — APPROVED** [review]
Can you PTAL @eddumelendez?

**Review by eddumelendez — APPROVED** [review]

**Pull request merged**